### PR TITLE
Ignore passing moves unless they make sense.

### PR DESCRIPTION
**Pull request by gcp**
See discussion in #2277.

### CODE REVIEW

File: src/UCTNode.cpp [hunk 1]
```diff
@@ -103,8 +103,24 @@ bool UCTNode::create_children(Network & network,
...
     }
-    nodelist.emplace_back(raw_netlist.policy_pass, FastBoard::PASS);
-    legal_sum += raw_netlist.policy_pass;
+
+    // Always try passes if we're not trying to be clever.
+    auto allow_pass = cfg_dumbpass;
+
+    // If we're clever, only try passing if we're winning on the
+    // net score and on the board count.
+    if (!allow_pass && (m_net_eval > 0.8f || nodelist.size() <= 10)) {
```
Comment (Ttl) [inline src/UCTNode.cpp]: `m_net_eval` is from black's point of view. It needs to be inverted if the player is white.

Comment (gcp) [reply src/UCTNode.cpp]: That's what the code above this already does.

Comment (Ttl) [reply src/UCTNode.cpp]: NN returns value from the point of the current player and the above code makes it black's view for the search. It needs to be inverted back for the current players view for this check.

**Comment by gcp** [comment]
Wasn't sure what to put the minimum legal moves at. It needs to "reasonably" cover all seki points on the board. 

If there are only eye-points left, then get_nopass_child will already prevent disaster at the root, but the engine could get pretty confused while searching ahead. So it tried to put it high enough, i.e. to count for like 7 or 8 groups and some seki points.

**Comment by Ttl** [comment]
This solves the issue and could be even better solution since it doesn't consider passing at all now. In handicap games with very low eval it did still consider passing with the other patch and could in some situations even play pass.

There could be potential issue where passing would win and it would be sensible to pass, but the NN doesn't realize it and doesn't give high enough value output. At least some of the earlier nets had issues with scoring close games, but I don't think the latest networks really make this mistake anymore. Maybe not worth worrying about?

**Comment by gcp** [comment]
>Maybe not worth worrying about?

It's tricky to do right. The logic here avoids counting in most cases. Even if you'd always count, you would still have the problem that the engine would start adding passes whenever the board is winning in Trump Taylor, which could happen in the middle of tight middlegames.

Note that we will pass anyway when we start running out of moves.

**Comment by alreadydone** [comment]
What about marking the pass move as invalid when it is selected and if the position is net-winning but score-losing (and if at least one visited sibling isn't marked invalid), like in https://github.com/leela-zero/leela-zero/blob/master/src/UCTSearch.cpp#L227-L237? This involve score checks but should reduce them to a minimum.

**Comment by gcp** [comment]
>and if at least one visited sibling isn't marked invalid

Sounds much more complicated. The only thing it would save is score checks if the score is >80% winrate, and I think once the score is there we don't care about the tiny performance impact - if any - any more.

**Pull request merged**